orttraining/orttraining/models/runner/training_runner.cc

```diff
@@ -495,28 +795,37 @@ Status TrainingRunner::Evaluate(InferenceSession& session, IDataLoader& data_loa
         num_batches * params_.batch_size);
   }
 
-  OrtValue loss_scale_val;
-  TrainingUtil::CreateCpuMLValue({1}, std::vector<float>{params_.loss_scale}, &loss_scale_val);
-
   RunOptions run_options;
-  run_options.only_execute_path_to_fetches = true;
   for (size_t batch_idx = 0; batch_idx < num_batches; ++batch_idx) {
-    std::vector<MLValue> feeds = test_data->GetKthBatch(params_.batch_size, current_batch);
-    if (loss_scaler_) {
-      feeds.push_back(loss_scale_val);
-    }
-    OrtValue lr_ort_val;
-    TrainingUtil::CreateCpuMLValue({1}, std::vector<float>{params_.lr_params.initial_lr}, &lr_ort_val);
-    feeds.push_back(lr_ort_val);
+    std::vector<std::string> feed_names;
+    std::vector<MLValue> feeds;
+    std::vector<std::string> fetch_names;
     std::vector<MLValue> fetches;
+
+    PrepareFeedNamesAndFeeds(EvaluateStep,
+                             data_loader,
+                             *test_data,
+                             nullptr,
+                             batch_idx,
+                             feed_names,
+                             feeds);
+
+    PrepareFetchNamesAndFetches(EvaluateStep,
+                                fetch_names,
+                                fetches);
+
     ORT_RETURN_IF_ERROR(session.Run(run_options,
                                     feed_names,
                                     feeds,
-                                    params_.fetch_names,
+                                    fetch_names,
                                     &fetches));
 
+    // Assume that user-specified fetches are avaliable only on the last pipeline stage. 
+    const bool session_can_see_loss = params_.pipeline_parallel_size == 1 ||
```

Comment:
>params_.pipeline_parallel_size == 1 [](start = 38, length = 35)

this is implied in pipeline_context_.pipeline_stage_id == params_.pipeline_parallel_size - 1;

Reply:
You're right. I also change its comment to
```
    // Assume that user-specified fetches are avaliable only on the last pipeline stage. 
    // When there is no pipeline, all pipeline_context_.pipeline_stage_id should be 0 and
    // params_.pipeline_parallel_size is 1. Thus, the following condition is always true if there
    // is no pipeline.
    const bool session_can_see_loss = pipeline_context_.pipeline_stage_id == params_.pipeline_parallel_size - 1;
```